### PR TITLE
Feature/quaternion

### PR DESCRIPTION
**Pull request by mattngc**
Adds support for Quaternions as per #264.

### CODE REVIEW

File: raysect/core/math/quaternion.pyx [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+        return self.mul_scalar(n)
+
+    cpdef Quaternion copy(self):
+        """Returns a copy of this quaternion."""
+
+        return new_quaternion(self.s, self.x, self.y, self.z)
+
+    cpdef AffineMatrix3D to_transform(self):
+        """
+        Transforms the vector with the supplied AffineMatrix3D.
```
Comment (CnlPepper) [inline raysect/core/math/quaternion.pyx]: Docstring is for the vector.transform, needs replacing. I think this method should also be called as_matrix().

File: raysect/core/math/quaternion.pyx [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+
+        self.s = s
+        self.x = x
+        self.y = y
+        self.z = z
+
+    def __repr__(self):
+        """Returns a string representation of the Quaternion object."""
+
+        return "Quaternion(" + str(self.s) + ", <" + str(self.x) + ", " + str(self.y) + ", " + str(self.z) + ">)"
```
Comment (CnlPepper) [inline raysect/core/math/quaternion.pyx]: Violates intention of repr. It should either generate a string that can be copy-pasted to reinstance the object or be an alternate, non-code representation enclosed in <blah>. This should return: Quarternion(1, 0, 0, 1)  and not Quarternion(1, <0, 0, 1>) as this is invalid python.

Comment (CnlPepper) [reply raysect/core/math/quaternion.pyx]: This will mean fixing the docstring examples too.

File: raysect/core/math/quaternion.pyx [hunk 3]
```diff
@@ -0,0 +1,556 @@
+
+import numbers
+cimport cython
+from libc.math cimport sqrt, sin, cos, M_PI
+
+from raysect.core.math.affinematrix cimport new_affinematrix3d
+
+
+cdef double DEG2RAD = 2*M_PI / 360
```
Comment (CnlPepper) [inline raysect/core/math/quaternion.pyx]: Copy the DEF lines from math/transform.pyx, they are precalculated.

I'll make a central constants file at some point.

File: raysect/core/math/quaternion.pyx [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+        return new_quaternion(ns, nx, ny, nz)
+
+    cdef Quaternion mul_scalar(self, double d):
+
+        cdef Quaternion q = self
+
+        return new_quaternion(d * q.s, d * q.x, d * q.y, d * q.z)
+
+    @cython.cdivision(True)
+    cpdef Quaternion inv(self):
```
Comment (CnlPepper) [inline raysect/core/math/quaternion.pyx]: Please use inverse() not inv(), as per AffineMatrix3D: cpdef AffineMatrix3D inverse(self)

File: raysect/core/math/quaternion.pyx [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+        m30 = m31 = m32 = 0.0
+        m33 = 1.0
+
+        return new_affinematrix3d(m00, m01, m02, m03,
+                                  m10, m11, m12, m13,
+                                  m20, m21, m22, m23,
+                                  m30, m31, m32, m33)
+
+    @classmethod
+    def from_axis_angle(cls, axis, angle):
```
Comment (CnlPepper) [inline raysect/core/math/quaternion.pyx]: Missing object types

Comment (CnlPepper) [reply raysect/core/math/quaternion.pyx]: if the arg types are set, you don't need the isinstance tests.

File: raysect/core/math/quaternion.pyx [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+
+        qs = cos(theta_2)
+        qx = axis.x * sin(theta_2)
+        qy = axis.y * sin(theta_2)
+        qz = axis.z * sin(theta_2)
+
+        return new_quaternion(qs, qx, qy, qz)
+
+
+cpdef Quaternion matrix2quaternion(AffineMatrix3D matrix):
```
Comment (CnlPepper) [inline raysect/core/math/quaternion.pyx]: I hate this function name! Will think of an alternative.... probably just to_quarternion().

Alternately make this a class method on Quarternion eg: Quarternion.from_matrix() which has nice symmetry with to_matrix.

File: raysect/core/math/quaternion.pyx (hunk 1 above)
Comment (CnlPepper) [reply raysect/core/math/quaternion.pyx]: ...or to_matrix() to be consistent with from_matrix() suggestion below.

Comment (mattngc) [reply raysect/core/math/quaternion.pyx]: Resolved in a4d9d8a

File: raysect/core/math/quaternion.pyx (hunk 2 above)
Comment (mattngc) [reply raysect/core/math/quaternion.pyx]: Resolved in a4d9d8a

File: raysect/core/math/quaternion.pyx (hunk 3 above)
Comment (mattngc) [reply raysect/core/math/quaternion.pyx]: Resolved in a4d9d8a

File: raysect/core/math/quaternion.pyx (hunk 4 above)
Comment (mattngc) [reply raysect/core/math/quaternion.pyx]: Resolved in a4d9d8a

File: raysect/core/math/quaternion.pyx (hunk 5 above)
Comment (mattngc) [reply raysect/core/math/quaternion.pyx]: Resolved in a4d9d8a

File: raysect/core/math/quaternion.pyx (hunk 6 above)
Comment (mattngc) [reply raysect/core/math/quaternion.pyx]: Resolved in a4d9d8a

**Comment by mattngc** [comment]
Addressed all comments and ready for re-review, @CnlPepper.

**Pull request merged**